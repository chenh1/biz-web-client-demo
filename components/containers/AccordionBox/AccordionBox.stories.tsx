import type { AccordionBoxProps } from './AccordionBox';
import React from 'react';
import { AccordionBox } from './AccordionBox';
import { Paragraph } from '../../core/Paragraph/Paragraph';
import { Heading } from '../../core/Heading/Heading';
import { Button } from '../../core/Button/Button';
import { Box } from '../Box/Box';

type Args = {
  args: AccordionBoxProps
}

const Template = (args: AccordionBoxProps) => (
  <AccordionBox className={"max-w-xs"} {...args}>
    <Box padding="0" gap="sm">
      <Heading type="h3">Heading text</Heading>
      <Paragraph>This is some text for a sample paragraph element</Paragraph>
      <Button label="CLICK ME" primary/>
    </Box>
  </AccordionBox>
)

export default {
  component: Template
}

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const AccordionBoxDefault: Args = { args: 
  { label: "CLICK ME" } 
};

