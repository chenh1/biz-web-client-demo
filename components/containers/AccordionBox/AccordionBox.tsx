import React, { useState, useEffect, useRef, ReactNode } from 'react';
import { Box } from '../Box/Box';

export type AccordionBoxProps = {
  className?: string;
  padding?: 'sm' | 'md' | 'lg'; // Adjust these values based on your padding options
  label: ReactNode;
  children?: ReactNode;
  iconHide?: ReactNode;
  iconShow?: ReactNode;
}

export const AccordionBox: React.FC<AccordionBoxProps> = ({
  className = "",
  padding = "sm",
  label,
  children,
  iconHide,
  iconShow
}) => {
  const [showContent, setShowContent] = useState(false);
  const [hiddenMaxHeightClass, setHiddenMaxHeightClass] = useState("0"); // [px
  const labelBarRef = useRef<HTMLDivElement | null>(null);

  const toggleShowContent = () => {
    if (labelBarRef.current) {
      if (showContent) {
        setHiddenMaxHeightClass(`max-h-[${labelBarRef.current.offsetHeight}px]`);
      } else {
        setHiddenMaxHeightClass(`max-h-full`);
      }
    }
    setShowContent(!showContent);
  };

  useEffect(() => {
    if (labelBarRef.current) {
      setHiddenMaxHeightClass(`max-h-[${labelBarRef.current.offsetHeight}px]`);
    }
  }, []);

  return (
    <Box padding={padding} hasShadow isRounded className={`${className} box-content duration-300 ease-in-out ${hiddenMaxHeightClass}`}>
      <div data-testid="accordionLabelClick" ref={labelBarRef} className="cursor-pointer" onClick={toggleShowContent}>
        {label}
        {showContent ? iconHide : iconShow}
      </div>
      <div className={`relative overflow-hidden duration-300 ease-in-out ${showContent ? 'h-full' : 'h-0'}`}>
        <div className={`duration-300 ease-in-out ${showContent ? 'translate-y-0' : '-translate-y-full'}`}>
          {children}
        </div>
      </div>
    </Box>
  );
};