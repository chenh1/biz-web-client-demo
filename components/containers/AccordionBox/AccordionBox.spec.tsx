import { AccordionBox } from "./AccordionBox";
import { fireEvent, render } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window


it('renders closed accordion', async () => {
  const { getByTestId, asFragment } = render(
    <AccordionBox label={<h3>Label</h3>} />
  );

  expect(asFragment()).toMatchSnapshot();
});

it('renders opened accordion once label is clicked', async () => {
  const { getByTestId, asFragment } = render(
    <AccordionBox label={<h3>Label</h3>} />
  );

  fireEvent.click(getByTestId("accordionLabelClick"))

  expect(asFragment()).toMatchSnapshot();
});