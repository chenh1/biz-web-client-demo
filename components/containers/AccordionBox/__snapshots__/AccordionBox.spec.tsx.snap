// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders closed accordion 1`] = `
<DocumentFragment>
  <div
    class=" grid rounded-md gap-4 shadow-lg p-4  box-content duration-300 ease-in-out max-h-[0px]"
  >
    <div
      class="cursor-pointer"
      data-testid="accordionLabelClick"
    >
      <h3>
        Label
      </h3>
    </div>
    <div
      class="relative overflow-hidden duration-300 ease-in-out h-0"
    >
      <div
        class="duration-300 ease-in-out -translate-y-full"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders opened accordion once label is clicked 1`] = `
<DocumentFragment>
  <div
    class=" grid rounded-md gap-4 shadow-lg p-4  box-content duration-300 ease-in-out max-h-full"
  >
    <div
      class="cursor-pointer"
      data-testid="accordionLabelClick"
    >
      <h3>
        Label
      </h3>
    </div>
    <div
      class="relative overflow-hidden duration-300 ease-in-out h-full"
    >
      <div
        class="duration-300 ease-in-out translate-y-0"
      />
    </div>
  </div>
</DocumentFragment>
`;
